### PR TITLE
CPR-154

### PR DESCRIPTION
**Pull request by vb-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/matcher/PrisonerMatcher.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package uk.gov.justice.digital.hmpps.personrecord.service.matcher
+
+import uk.gov.justice.digital.hmpps.personrecord.client.model.Prisoner
+import uk.gov.justice.digital.hmpps.personrecord.model.Person
+import uk.gov.justice.digital.hmpps.personrecord.validate.PNCIdentifier
+
+class PrisonerMatcher(val prisonerDetails: List<Prisoner>?, val person: Person) {
```
Comment (vb-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/matcher/PrisonerMatcher.kt:7]: There are some subtle variations when compared to OffenderMatcher so keep this for now, Once we have all the matchers then will work out the best way to minimise the duplication.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PrisonerService.kt [hunk 2]
```diff
@@ -21,87 +21,79 @@ class PrisonerService(
...
-        telemetryService.trackEvent(
-          TelemetryEventType.NOMIS_NO_MATCH_FOUND,
-          mapOf("UUID" to personEntity.personId.toString(), "PNC" to person.otherIdentifiers?.pncIdentifier?.pncId),
-        )
+      val prisonerMatcher = getPrisonerMatcher(person)
+      when {
+        prisonerMatcher.isPncDoesNotMatch() -> pncDoesNotMatch(prisonerMatcher, personEntity, person)
+        prisonerMatcher.isExactMatch() -> exactMatchFound(prisonerMatcher, personEntity, person)
+        prisonerMatcher.isPartialMatch() -> partialMatchFound(personEntity, person)
+        else -> noRecordsFound(personEntity, person)
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PrisonerService.kt:37]: ❤️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/matcher/PrisonerMatcher.kt (hunk 1 above)
Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/matcher/PrisonerMatcher.kt:7]: 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PrisonerServiceTest.kt [hunk 3]
```diff
@@ -53,12 +53,12 @@ class PrisonerServiceTest {
       givenName = "John",
       familyName = "Doe",
       dateOfBirth = prisonerDOB,
-      otherIdentifiers = OtherIdentifiers(pncIdentifier = PNCIdentifier("PNC123")),
+      otherIdentifiers = OtherIdentifiers(pncIdentifier = PNCIdentifier("PNC123"), prisonNumber = "12345"),
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PrisonerServiceTest.kt:56]: I know it is not the focus of this change, but it is a shame we still have so many invalid PNCs knocking about in tests

**Review by markreesmoj — APPROVED** [review]
👍

**Pull request merged**